File: android/src/main/java/io/github/phiresky/wayland_android/SetupOverlay.java
package io.github.phiresky.wayland_android;

import android.app.Activity;
import android.graphics.PixelFormat;
import android.graphics.Typeface;
import android.os.Handler;
import android.os.Looper;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.WindowManager;
import android.widget.ScrollView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Adds a full-screen log overlay on top of the NativeActivity's EGL surface
 * during first-run proot setup. Uses WindowManager to place the view above
 * the native rendering surface.
 */
public class SetupOverlay {
    private static final List<String> sLines =
            Collections.synchronizedList(new ArrayList<>());
    private static volatile TextView sLogView;
    private static volatile ScrollView sScrollView;
    private static Handler sHandler;
    private static int sLastFlushed = 0;

    /** Called from native code via JNI. Adds the overlay on the UI thread. */
    public static void show(Activity activity) {
        sLines.clear();
        sLastFlushed = 0;
        activity.runOnUiThread(() -> {
            ScrollView scrollView = new ScrollView(activity);
            TextView logView = new TextView(activity);
            logView.setTextColor(0xFFCCCCCC);
            logView.setBackgroundColor(0xFF111111);
            logView.setTypeface(Typeface.MONOSPACE);
            logView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 11);
            logView.setPadding(32, 48, 32, 32);
            logView.setText("=== Setup starting ===\n");
            scrollView.addView(logView);

            WindowManager.LayoutParams params = new WindowManager.LayoutParams(
                    WindowManager.LayoutParams.MATCH_PARENT,
                    WindowManager.LayoutParams.MATCH_PARENT,
                    WindowManager.LayoutParams.TYPE_APPLICATION_PANEL,
                    WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                            | WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN,
                    PixelFormat.OPAQUE);
            params.gravity = Gravity.FILL;
            params.token = activity.getWindow().getDecorView().getWindowToken();

            activity.getWindowManager().addView(scrollView, params);

            sLogView = logView;
            sScrollView = scrollView;
            sHandler = new Handler(Looper.getMainLooper());
            sHandler.post(SetupOverlay::flushLines);
        });
    }

    private static void flushLines() {
        TextView logView = sLogView;
        ScrollView scrollView = sScrollView;
        if (logView == null || scrollView == null) return;

        synchronized (sLines) {
            for (int i = sLastFlushed; i < sLines.size(); i++) {
                logView.append(sLines.get(i) + "\n");
            }
            sLastFlushed = sLines.size();
        }
        scrollView.post(() -> scrollView.fullScroll(ScrollView.FOCUS_DOWN));
        if (sHandler != null) {
            sHandler.postDelayed(SetupOverlay::flushLines, 100);
        }
    }

    /** Called from native code via JNI to append a log line. Thread-safe. */
    public static void appendLog(String line) {
        sLines.add(line);
    }

    /** Called from native code via JNI when setup is complete. Removes the overlay. */
    public static void hide(Activity activity) {
        if (sHandler != null) {
            sHandler.removeCallbacksAndMessages(null);
        }
        activity.runOnUiThread(() -> {
            // Flush remaining lines
            flushLines();
            if (sHandler != null) {
                sHandler.removeCallbacksAndMessages(null);
            }
            if (sScrollView != null) {
                try {
                    activity.getWindowManager().removeView(sScrollView);
                } catch (Exception e) {
                    // View might not be attached
                }
            }
            sLogView = null;
            sScrollView = null;
            sHandler = null;
        });
    }
}

File: android/src/main/java/io/github/phiresky/wayland_android/SetupOverlay.kt
package io.github.phiresky.wayland_android

import android.app.Activity
import android.graphics.PixelFormat
import android.graphics.Typeface
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.Gravity
import android.view.WindowManager
import android.widget.ScrollView
import android.widget.TextView
import java.util.Collections

/**
 * Adds a full-screen log overlay on top of the MainActivity's window
 * during first-run proot setup. Uses WindowManager to place the view above
 * the native rendering surface.
 */
object SetupOverlay {

    private val lines: MutableList<String> = Collections.synchronizedList(mutableListOf())

    @Volatile
    private var logView: TextView? = null

    @Volatile
    private var scrollView: ScrollView? = null
    private var handler: Handler? = null
    private var lastFlushed = 0

    /** Called from native code via JNI. Adds the overlay on the UI thread. */
    @JvmStatic
    fun show(activity: Activity) {
        lines.clear()
        lastFlushed = 0
        activity.runOnUiThread {
            val sv = ScrollView(activity)
            val lv = TextView(activity).apply {
                setTextColor(0xFFCCCCCC.toInt())
                setBackgroundColor(0xFF111111.toInt())
                typeface = Typeface.MONOSPACE
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 11f)
                setPadding(32, 48, 32, 32)
                text = "=== Setup starting ===\n"
            }
            sv.addView(lv)

            val params = WindowManager.LayoutParams(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.TYPE_APPLICATION_PANEL,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                    or WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN,
                PixelFormat.OPAQUE
            ).apply {
                gravity = Gravity.FILL
                token = activity.window.decorView.windowToken
            }

            activity.windowManager.addView(sv, params)

            logView = lv
            scrollView = sv
            handler = Handler(Looper.getMainLooper()).also {
                it.post(::flushLines)
            }
        }
    }

    private fun flushLines() {
        val lv = logView ?: return
        val sv = scrollView ?: return

        synchronized(lines) {
            for (i in lastFlushed until lines.size) {
                lv.append(lines[i] + "\n")
            }
            lastFlushed = lines.size
        }
        sv.post { sv.fullScroll(ScrollView.FOCUS_DOWN) }
        handler?.postDelayed(::flushLines, 100)
    }

    /** Called from native code via JNI to append a log line. Thread-safe. */
    @JvmStatic
    fun appendLog(line: String) {
        lines.add(line)
    }

    /** Called from native code via JNI when setup is complete. Removes the overlay. */
    @JvmStatic
    fun hide(activity: Activity) {
        activity.runOnUiThread {
            // Flush remaining lines before removing the overlay.
            flushLines()
            handler?.removeCallbacksAndMessages(null)
            scrollView?.let { sv ->
                try {
                    activity.windowManager.removeView(sv)
                } catch (_: Exception) {
                    // View might not be attached
                }
            }
            logView = null
            scrollView = null
            handler = null
        }
    }
}
